ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/TestExecutor.java

```diff
@@ -0,0 +1,244 @@
+      final TestCase testCase,
+      final ServiceContext serviceContext,
+      final KsqlEngine ksqlEngine,
+      final KsqlConfig ksqlConfig) {
+    final Map<String, String> persistedConfigs = testCase.persistedProperties();
+    final KsqlConfig maybeUpdatedConfigs = persistedConfigs.isEmpty() ? ksqlConfig :
+        ksqlConfig.overrideBreakingConfigsWithOriginalValues(persistedConfigs);
+
+    final List<PersistentQueryMetadata> queryMetadataList = buildQueries(
+        testCase, serviceContext, ksqlEngine, maybeUpdatedConfigs);
+    final List<TopologyTestDriverContainer> topologyTestDrivers = new ArrayList<>();
+    for (final PersistentQueryMetadata persistentQueryMetadata: queryMetadataList) {
+      final Properties streamsProperties = new Properties();
+      streamsProperties.putAll(persistentQueryMetadata.getStreamsProperties());
+      final TopologyTestDriver topologyTestDriver = new TopologyTestDriver(
+          persistentQueryMetadata.getTopology(),
+          streamsProperties,
+          0);
+      topologyTestDrivers.add(TopologyTestDriverContainer.of(
+          topologyTestDriver,
+          persistentQueryMetadata.getSourceNames()
+              .stream()
+              .map(s -> ksqlEngine.getMetaStore().getSource(s).getKsqlTopic())
+              .collect(Collectors.toSet()),
+          ksqlEngine.getMetaStore().getSource(persistentQueryMetadata.getSinkNames()
+              .iterator().next()).getKsqlTopic()
+      ));
+    }
+    return topologyTestDrivers;
+  }
+
+  private static List<PersistentQueryMetadata> buildQueries(
+      final TestCase testCase,
+      final ServiceContext serviceContext,
+      final KsqlEngine ksqlEngine,
+      final KsqlConfig ksqlConfig
+  ) {
+    testCase.initializeTopics(
+        serviceContext.getTopicClient(),
+        serviceContext.getSchemaRegistryClient());
+
+    final String sql = testCase.statements().stream()
+        .collect(Collectors.joining(System.lineSeparator()));
+
+    final List<PersistentQueryMetadata> queries = execute(
+        ksqlEngine,
+        sql,
+        ksqlConfig,
+        testCase.properties(),
+        Optional.of(serviceContext.getSchemaRegistryClient())
+    );
+
+    assertThat("test did not generate any queries.", queries.isEmpty(), is(false));
+    return queries;
+  }
+
+  /**
+   * @param srClient if supplied, then schemas can be inferred from the schema registry.
+   */
+  private static List<PersistentQueryMetadata> execute(
+      final KsqlEngine engine,
+      final String sql,
+      final KsqlConfig ksqlConfig,
+      final Map<String, Object> overriddenProperties,
+      final Optional<SchemaRegistryClient> srClient
+  ) {
+    final List<ParsedStatement> statements = engine.parse(sql);
+
+    final Optional<DefaultSchemaInjector> schemaInjector = srClient
+        .map(SchemaRegistryTopicSchemaSupplier::new)
+        .map(DefaultSchemaInjector::new);
+
+    final KsqlExecutionContext sandbox = engine.createSandbox();
```

Comment:
why do we need to do validation in the test environment? isn't the whole point to validate with the results?

Reply:
This validates the whole pipeline, but we may not validate all the results.

Reply:
sure. but this happens anyway when we execute it in the statement below, no? see #2803

Reply:
I see what you mean, good point. Removed this.